kernel/aster-nix/src/time/wait.rs

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::time::Duration;
+
+use aster_frame::sync::{WaitQueue, Waiter};
+
+use super::clock::JIFFIES_TIMER_MANAGER;
+
+/// A trait that provide the timeout related function for [`WaitQueue`]`.
+pub trait WaitTimeout {
+    /// Wait until some condition returns `Some(_)`, or a given timeout is reached. If
+    /// the condition does not becomes `Some(_)` before the timeout is reached, the
+    /// function will return `None`.
+    fn wait_until_or_timeout<F, R>(&self, cond: F, timeout: &Duration) -> Option<R>
+    where
+        F: FnMut() -> Option<R>;
+}
+
+impl WaitTimeout for WaitQueue {
+    fn wait_until_or_timeout<F, R>(&self, mut cond: F, timeout: &Duration) -> Option<R>
+    where
+        F: FnMut() -> Option<R>,
+    {
+        if let Some(res) = cond() {
+            return Some(res);
+        }
+
+        let (waiter, waker) = Waiter::new_pair();
+        let wake_up = {
+            let waker = waker.clone();
+            move || {
+                waker.wake_up();
+            }
+        };
+
+        let jiffies_timer = JIFFIES_TIMER_MANAGER.get().unwrap().create_timer(wake_up);
+        jiffies_timer.set_timer(*timeout);
```

Comment:
> What's the advantage of calling IntervalTimer::set_timer instead of directly registering a TimerCallback?

This is because `TimerCallback` is not a timer. It does not have a method to read time and should be managed by `TimerManager`. Well I now think `TimerCallback` should not be public and I will correct it.

> When I first saw the name IntervalTimer, I thought it's a periodic timer. But this understanding is wrong.

`IntervalTimer` is a timer that can be periodic. It the `interval` field in it is 0 it will be a one shot timer.

> So why not making IntervalTimer only a periodic timer

This is because some syscalls will operate a specific timer, and can change it to a periodic timer or an one shot timer. So I think it is better for the `IntervalTimer` to have both abilities.

> Are you sure that jiffies_timer.clear(); is better than jiffies_timer.cancel();?

I agree, naming `cancel` will be better.

Reply:
> This is because some syscalls will operate a specific timer, and can change it to a periodic timer or an one shot timer.

Thanks for the explanation. It makes sense to me.

Can you also rename `set_timer` to `set_timeout`? The former naming is quite confusing. After all, `IntervalTimer` itself is a timer, so the `set_timer` method sounds like a `set` method which does not make any sense.

Besides, I think `IntervalTimer` should take the interval/timeout in its constructors, e.g. having a `new_oneshot` and a `new_periodic` methods instead of a `new` method.

Again, I don't have time to take a comprehensive look at these APIs. I have only checked the APIs used by `WaitQueue`. Perhaps there is still room to improve other APIs together. (CCing @StevenJiang1110 since I will _not_ review this PR other than the part related to `WaitQueue`.)

Reply:
> Can you also rename set_timer to set_timeout? The former naming is quite confusing. 

OK, `set_timeout` seems better.

> How do you like just naming the `IntervalTimer` as `Timer`?

The reason I named it `IntervalTimer` is to be consistent with Linux, as Linux uses the term `itimer` to refer to timers within the system. Therefore, I think the name `itimer` (i.e. `IntervalTimer`) might be more widely recognized. In Linux, `itimer` inherently signifies a timer with periodic capabilities that can be activated and can also switch between oneshot and periodic states at any time via syscalls. Consequently, it might not be feasible to implement separate constructor functions for periodic and oneshot scenarios, because this state is a mutable attribute of itimer.

Of course, I understand that this might lead to some ambiguity. Perhaps I could define a separate oneshot timer, but that seems to add complexity and redundancy to the code. Maybe I can resolve these ambiguities through documentation.

Reply:
> The reason I named it `IntervalTimer` is to be consistent with Linux

> Consequently, it might not be feasible to implement separate constructor functions for periodic and oneshot scenarios, because this state is a mutable attribute of itimer

OK. Thanks for your explanation. Given the widely used Linux definitions (with which I'm not very familiar), the current API design may be acceptable, so I won't insist on these points.